### PR TITLE
House price prediction examples

### PR DESCRIPTION
**Pull request by cfascina**
## Impact and Context
Example notebooks with house price prediction using Gradient Boosting Regressor (scikit-learn) and LightGBM.

## Testing
- [ ] Unit test
- [ ] Deployed to dev env
- [X] Other (explain) 

## Reverting
- [ ] Contains Migration - _Do Not Revert_

### CODE REVIEW

File: deployment/sklearn/sklearn-housing-gb-regression.ipynb [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+        "        self.model = cloudpickle.load(open(artifacts['serialized_model'], 'rb'))\n",
+        "        \n",
+        "    @verify_io\n",
+        "    def predict(self, data):\n",
+        "        results = []\n",
+        "\n",
+        "        for item in data:\n",
+        "            results.append(self.model.predict(item).tolist())\n",
+        "        \n",
+        "        return results"
```
Comment (liuverta) [inline deployment/sklearn/sklearn-housing-gb-regression.ipynb:307]: See my comment in the other notebook!

File: deployment/sklearn/sklearn-housing-gb-regression.ipynb [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+      ]
```
Comment (liuverta) [inline deployment/sklearn/sklearn-housing-gb-regression.ipynb:308]: Also add `def example()` to this model if you can!

File: deployment/lightgbm/lightgbm-house-price-prediction.ipynb [hunk 3]
```diff
@@ -0,0 +1,488 @@
...
+        },
+        "id": "7TIE5fwEJ7OJ",
+        "outputId": "e60f84c5-5019-41a2-bc83-dee21b2a87e5"
+      },
+      "outputs": [],
+      "source": [
+        "gbm = lgb.train(\n",
+        "    params,\n",
+        "    lgb_train,\n",
+        "    num_boost_round = 20,\n",
```
[the remark this answers is not in the document]
Comment (cfascina) [reply deployment/lightgbm/lightgbm-house-price-prediction.ipynb:225]: Done, metrics improved a lot with `num_boost_round = 200`.

File: deployment/lightgbm/lightgbm-house-price-prediction.ipynb [hunk 4]
```diff
@@ -0,0 +1,488 @@
...
+        "id": "85cbJc3VRAkU",
+        "outputId": "8db5ea0a-c55b-4194-d1cc-63eed5b098e7"
+      },
+      "outputs": [],
+      "source": [
+        "model = registered_model.create_standard_model(\n",
+        "    name = 'v0',\n",
+        "    model_cls = Predictor,\n",
+        "    environment = Python(requirements=['scikit-learn', 'lightgbm']),\n",
+        "    model_api = ModelAPI(X_train, pd.DataFrame(data = y_train.tolist(), columns = ['PricePrediction'])),\n",
```
[the remark this answers is not in the document]
Comment (cfascina) [reply deployment/lightgbm/lightgbm-house-price-prediction.ipynb:407]: Done.

File: deployment/lightgbm/lightgbm-house-price-prediction.ipynb [hunk 5]
```diff
@@ -0,0 +1,488 @@
...
+        "        self.model = cloudpickle.load(open(artifacts['serialized_model'], 'rb'))\n",
+        "    \n",
+        "    @verify_io\n",
+        "    def predict(self, data):\n",
+        "        results = []\n",
+        "\n",
+        "        for item in data:\n",
+        "            results.append(self.model.predict(item).tolist())\n",
+        "        \n",
+        "        return results\n",
```
[the remark this answers is not in the document]
Comment (cfascina) [reply deployment/lightgbm/lightgbm-house-price-prediction.ipynb:349]: Indeed. Done.

File: deployment/sklearn/sklearn-housing-gb-regression.ipynb (hunk 2 above)
Comment (cfascina) [reply deployment/sklearn/sklearn-housing-gb-regression.ipynb:308]: Done.

File: deployment/sklearn/sklearn-housing-gb-regression.ipynb (hunk 1 above)
Comment (cfascina) [reply deployment/sklearn/sklearn-housing-gb-regression.ipynb:307]: Done.

File: deployment/lightgbm/lightgbm-house-price-prediction.ipynb [hunk 6]
```diff
@@ -0,0 +1,566 @@
...
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "N4GOi5ucWywL"
+      },
+      "outputs": [],
+      "source": [
+        "sample_data = X_train.values.tolist()[:5]"
```
Comment (liuverta) [inline deployment/lightgbm/lightgbm-house-price-prediction.ipynb:477]: I think we should use `X_test` here, because it is available in the notebook and the model is already trained (and then maybe we also don't need `cols` below anymore)

File: deployment/sklearn/sklearn-housing-regression-gbr.ipynb [hunk 7]
```diff
@@ -0,0 +1,522 @@
...
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "dqGPqltGo7Pd"
+      },
+      "outputs": [],
+      "source": [
+        "sample_data = X_train.values.tolist()[:5]"
```
Comment (liuverta) [inline deployment/sklearn/sklearn-housing-regression-gbr.ipynb:437]: See my comment in the other notebook about `X_test`.

File: deployment/lightgbm/lightgbm-house-price-prediction.ipynb (hunk 5 above)
[the remark this answers is not in the document]
Comment (liuverta) [reply deployment/lightgbm/lightgbm-house-price-prediction.ipynb]: Could you describe

> otherwise the output in WebApp is not filled

? I think I am able to see API in the WebApp when I use `ModelAPI(X_train, y_train)`, and `y_train` is more accurate than `y_train.to_frame()`.

[the remark this answers is not in the document]
Comment (cfascina) [reply deployment/lightgbm/lightgbm-house-price-prediction.ipynb]: Below are screenshots of the WebApp Input & Output Section for each case.

Using `ModelAPI(X_train, y_train.to_frame(name='PricePrediction'))`:
Captura de tela de 2023-01-03 11-33-47

Using `ModelAPI(X_train, y_train)`:
Captura de tela de 2023-01-03 11-36-49

Notice that the Output section is not filled when `y_train` is not converted to a Dataframe.

How should we proceed?

File: deployment/lightgbm/lightgbm-house-price-prediction.ipynb (hunk 6 above)
Comment (cfascina) [reply deployment/lightgbm/lightgbm-house-price-prediction.ipynb:477]: Changed from `X_train` to `X_test` :+1:

I still need to specify the columns, as `sample_data = X_test.values.tolist()[:5]` returns a list containing only the DataFrame values (excluding its columns names).

Even so, the code got cleaner with: `df = pd.DataFrame(data=sample_data, columns=X_test.columns)`
The cell that was creating a list with columns names is not necessary.

File: deployment/sklearn/sklearn-housing-regression-gbr.ipynb (hunk 7 above)
Comment (cfascina) [reply deployment/sklearn/sklearn-housing-regression-gbr.ipynb]: Done.

File: deployment/sklearn/sklearn-housing-regression-gbr.ipynb [hunk 8]
```diff
@@ -0,0 +1,522 @@
...
+        "    @verify_io\n",
+        "    def predict(self, data):\n",
+        "        return self.model.predict(data).tolist()\n",
+        "\n",
+        "    def describe(self):\n",
+        "        return {\n",
+        "            'method': 'predict',\n",
+        "            'args': f\"{self.example()}\",\n",
+        "            'returns': 'PricePrediction',\n",
+        "            'description': 'Predicts house prices based on LightGMB trained model.',\n",
```
Comment (liuverta) [inline deployment/sklearn/sklearn-housing-regression-gbr.ipynb:318]:
```suggestion
        "            'description': 'Predicts house prices based on scikit-learn's GradientBoostingRegressor trained model.',\n",
```

**Review by liuverta — CHANGES_REQUESTED** [review]

File: deployment/lightgbm/lightgbm-house-price-prediction.ipynb [hunk 9]
```diff
@@ -0,0 +1,562 @@
...
+        "    \n",
+        "    @verify_io\n",
+        "    def predict(self, data):\n",
+        "        return self.model.predict(data).tolist()\n",
+        "\n",
+        "    def describe(self):\n",
+        "        return {\n",
+        "            'method': 'predict',\n",
+        "            'args': f\"{self.example()}\",\n",
+        "            'returns': 'PricePrediction',\n",
```
[the remark this answers is not in the document]
Comment (cfascina) [reply deployment/lightgbm/lightgbm-house-price-prediction.ipynb:357]: Done.

File: deployment/sklearn/sklearn-housing-regression-gbr.ipynb [hunk 10]
```diff
@@ -0,0 +1,522 @@
...
+        "        \n",
+        "    @verify_io\n",
+        "    def predict(self, data):\n",
+        "        return self.model.predict(data).tolist()\n",
+        "\n",
+        "    def describe(self):\n",
+        "        return {\n",
+        "            'method': 'predict',\n",
+        "            'args': f\"{self.example()}\",\n",
+        "            'returns': 'PricePrediction',\n",
```
[the remark this answers is not in the document]
Comment (cfascina) [reply deployment/sklearn/sklearn-housing-regression-gbr.ipynb:317]: Done.

File: deployment/lightgbm/lightgbm-house-price-prediction.ipynb [hunk 11]
```diff
@@ -0,0 +1,562 @@
...
+        "            'description': 'Predicts house prices based on LightGMB trained model.',\n",
```
[the remark this answers is not in the document]
Comment (cfascina) [reply deployment/lightgbm/lightgbm-house-price-prediction.ipynb:358]: Done.

File: deployment/sklearn/sklearn-housing-regression-gbr.ipynb (hunk 8 above)
Comment (cfascina) [reply deployment/sklearn/sklearn-housing-regression-gbr.ipynb:318]: Done.

**Review by ewagner-verta — APPROVED** [review]
Seems like all comments were addressed

**Review by liuverta — APPROVED** [review]
